### PR TITLE
Refactor Hazelcast AWS Plugin

### PR DESCRIPTION
**Pull request by leszko**
My **main goals** for this refactoring PR are to:
* **Untangle** DescribeInstances and Metadata API usages (independent class for each API)
* Make `AwsDescribeInstancesApi` usage **non-static**
* Make `ApiConfig` **immutable**
* Give the module general structure similar to [hazelcast-kubernetes](https://github.com/hazelcast/hazelcast-kubernetes), [hazelcast-azure](https://github.com/hazelcast/hazelcast-azure), and [hazelcast-gcp](https://github.com/hazelcast/hazelcast-gcp)
* **Restructure packages** (remove most packages, we don't need them in such a small module)
* Remove deprecated **TcpIpJoinerOverAWS** (it's not used anywhere anymore)

-----------------------------

This PR introduces **no functional changes** other than:
* Some API calls are executed during class creation (before, it was done lazily) - in the context of Hazelcast it doesn't matter much, but in general IMO fail fast is better
* Some minor logging changes

----------------------------

This PR is **not intended to fix all problems** of the `hazelcast-aws` module. To make PR small and to make the revert simpler in case of some issues, I will follow up with the next refactoring as separate PRs (though definitely smaller than this one). I see the next steps as:
* Fix **testing** (introduce WireMock) and **coupling** of classes: `AwsDescribeInstancesApi`, `CloudyUtility`, `EC2RequestSigner`, and `Constants`, `Filter`, `AwsURLEncoder`
* Introduce [RestClient ](https://github.com/hazelcast/hazelcast-azure/blob/master/src/main/java/com/hazelcast/azure/RestClient.java) for REST calls
* Fix ECS discovery (I believe it doesn't work, but I was afraid to remove it), I'll sort it out while working on Fargate

### CODE REVIEW

File: src/main/java/com/hazelcast/aws/AwsCredentials.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+package com.hazelcast.aws;
+
+import java.util.Objects;
+
+public final class AwsCredentials {
```
Comment (pivovarit) [inline src/main/java/com/hazelcast/aws/AwsCredentials.java:5]: Does this need to be public? it's used just internally, right?

File: src/main/java/com/hazelcast/aws/AwsDiscoveryStrategy.java [hunk 2]
```diff
@@ -50,100 +48,69 @@
...
+        } catch (IllegalArgumentException e) {
+            throw new InvalidConfigurationException("AWS configuration is not valid", e);
         }
     }
 
     /**
      * Returns port range from properties or default value if the property does not exist.
      * <p>
-     * Note that {@link AbstractDiscoveryStrategy#getOrDefault(PropertyDefinition, Comparable)} cannot be reused, since
+     * Note that {@code AbstractDiscoveryStrategy#getOrDefault(PropertyDefinition, Comparable)} cannot be reused, since
```
Comment (pivovarit) [inline src/main/java/com/hazelcast/aws/AwsDiscoveryStrategy.java:113]: Why? `@link` makes the reference clickable

File: src/main/java/com/hazelcast/aws/AwsProperties.java [hunk 3]
```diff
@@ -81,7 +81,7 @@
     CONNECTION_TIMEOUT_SECONDS("connection-timeout-seconds", INTEGER, true),
 
     /**
-     * Number of retries while connecting to AWS Services. Its default value is 10.
+     * Number of retries while connecting to AWS Services. Its default value is 3.
```
Comment (pivovarit) [inline src/main/java/com/hazelcast/aws/AwsProperties.java:84]: Are you changing the default value or the documentation was wrong?

File: src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java [hunk 4]
```diff
@@ -43,24 +41,25 @@
     private static final int DATE_LENGTH = 8;
     private static final int LAST_INDEX = 8;
 
-    private final AwsConfig config;
     private final String timestamp;
 
     private String service;
     private Map<String, String> attributes;
+    private String region;
```
Comment (pivovarit) [inline src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java:48]: If these are not mutable, let's make them final?

File: src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java [hunk 5]
```diff
@@ -192,7 +191,7 @@ private String sha256Hashhex(String in) {
     }
 
     public String createFormattedCredential() {
-        return config.getAccessKey() + '/' + timestamp.substring(0, LAST_INDEX) + '/' + config.getRegion() + '/'
+        return credentials.getAccessKey() + '/' + timestamp.substring(0, LAST_INDEX) + '/' + region + '/'
```
Comment (pivovarit) [inline src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java:194]: `String.format` all of it?

**Comment by pivovarit** [comment]
I would love to see internals hidden(package-private) as much as possible if you're collapsing the structure of the project already. The more we had, the less it is to worry about backward compatibility in long term

File: src/main/java/com/hazelcast/aws/AwsProperties.java (hunk 3 above)
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsProperties.java:84]: The documentation was wrong.

File: src/main/java/com/hazelcast/aws/AwsDiscoveryStrategy.java (hunk 2 above)
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsDiscoveryStrategy.java:113]: Right, reverted.

File: src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java (hunk 4 above)
Comment (leszko) [reply src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java:48]: Added.

File: src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java (hunk 5 above)
Comment (leszko) [reply src/main/java/com/hazelcast/aws/utility/EC2RequestSigner.java:194]: I'll address it in a separate PR while working on refactoring EC2RequestSigner.

**Comment by leszko** [comment]
> I would love to see internals hidden(package-private) as much as possible if you're collapsing the structure of the project already. The more we had, the less it is to worry about backward compatibility in the long term

I thought that I would finish restructuring packages in the next PR, but you inspired me and I just flattened the whole package structure here. Therefore, all scopes are package-private (or private).

File: src/main/java/com/hazelcast/aws/AwsCredentials.java (hunk 1 above)
Comment (leszko) [reply src/main/java/com/hazelcast/aws/AwsCredentials.java]: done

**Comment by leszko** [comment]
verify

**Pull request merged**